LLEXT dependencies: add support for linking between LLEXT objects

lyakh opened the pull request:
This adds support for exporting symbols between extensions while building dependencies based on such symbol references

lyakh committed:
llext: remove an unused variable

Remove a loop counter, that isn't actually used for anything.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: look for symbols in other LLEXT objects too

LLEXT objects can also export symbols for use by other such objects.
That means, that when linking an LLEXT object we have to look for
unresolved symbols in previously loaded objects too.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: export a symbol needed for immediate logging

Using the immediate logging option from LLEXT modules requires
one more symbol export.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: fix flag evaluation for section grouping

When deciding which sections to group together, only the low 3
section attribute bits are relevant, ignore the rest.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: add dependencies

When an LLEXT object uses symbols of another such object, it depends
on it. Such dependencies have to be tracked to prevent their
accidental unloading. Ideally we should be able to track arbitrary
numbers of such dependencies, but this is a bit difficult. In this
first implementation we use a fixed-size array, currently consisting
of 8 entries.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

faxe1008 commented:
@lyakh sorry for marking this as ready for review, I misclicked in the GitHub mobile app. Apologies  for the noise.

teburd commented on subsys/llext/llext_link.c:
```diff
@@ -64,6 +112,7 @@ static int llext_get_symbol(struct llext *ext, void *arg)
 
 	if (addr) {
 		se->addr = addr;
+		se->ext = ext;
```
a bit confusing, why is the extension assigned to the struct when its passed in as a parameter already

teburd commented on subsys/llext/llext_link.c:
```diff
@@ -52,7 +52,55 @@ static size_t llext_file_offset(struct llext_loader *ldr, size_t offset)
+}
+
 struct llext_sym_ext {
+	struct llext *ext;
```
This seems like the wrong place for this

teburd commented on subsys/llext/llext_link.c:
```diff
@@ -159,18 +208,23 @@ static void llext_link_plt(struct llext_loader *ldr, struct llext *ext,
 			if (!link_addr) {
+				/* Finally try any loaded tables */
 				struct llext_sym_ext se = {.sym = name};
 				ret = llext_iterate(llext_get_symbol, &se);
```
I see why we need ext added to the llext_sym_ext now, but this still seems like the wrong place for the added member.

mathieuchopstm replied:
+1

Would be OK to me if names were different (e.g., `struct llext_sym_ext` -> `struct global_sym_search_workbuf` and `llext_get_symbol` -> `search_symbol_by_name`); but really, I'd rather see this wrapped in a `int llext_search_sym(const void *name, struct llext **exporting_module)` function (which could even be made public? although IDK if it would be very useful...)

lyakh replied:
> I see why we need ext added to the llext_sym_ext now, but this still seems like the wrong place for the added member.

@teburd sorry, don't quite understand. Why is this a wrong place? This structure is only local for this sole iteration run. It's introduced in this very PR a few commits earlier. Why cannot I add members to it?

lyakh replied:
> +1
> 
> Would be OK to me if names were different (e.g., `struct llext_sym_ext` -> `struct global_sym_search_workbuf` and `llext_get_symbol` -> `search_symbol_by_name`); but really, I'd rather see this wrapped in a `int llext_search_sym(const void *name, struct llext **exporting_module)` function (which could even be made public? although IDK if it would be very useful...)

@mathieuchopstm I could make it a function, sure, probably `static` for now, but don't think that structure deserves any "global" designation in its name - it's perfectly local and confined to this single use only

mathieuchopstm replied:
The "global" in my suggestion was meant to say "global search (for symbols)", as in "search for symbols everywhere". A more appropriate suggestion name suggestion would have been `llext_sym_search_param`.

IMO, the `llext_` prefix in iterate callback is confusing (sounds like an API you could call); and the structure name is cryptic which makes it confusing as to what its role is.

I think `int llext_search_sym()` could be useful in some situations (as `llext_iterate()` currently is :smile:), but I'm completely fine with just renaming.

teburd commented on subsys/llext/llext_link.c:
```diff
@@ -159,18 +208,23 @@ static void llext_link_plt(struct llext_loader *ldr, struct llext *ext,
 				ret = llext_iterate(llext_get_symbol, &se);
 
 				link_addr = se.addr;
+				if (link_addr)
```
missing brackets, will fail misra compliance

teburd commented on subsys/llext/llext_link.c:
```diff
@@ -52,7 +52,55 @@ static size_t llext_file_offset(struct llext_loader *ldr, size_t offset)
+	unsigned int i;
+
+	for (i = 0; i < ARRAY_SIZE(ext->dependency); i++) {
+		if (ext->dependency[i] == dependency)
```
missing brackets, will fail misra compliance

teburd reviewed (commented):
Looks mostly reasonable to me, some oddities that are maybe caused by using llext_iterate when perhaps instead a different approach would avoid some strangeness

Should fix the formatting issues otherwise misra checks will fail and someone just fixed many issues in the tree related to misra styling.

mathieuchopstm commented on subsys/llext/llext_load.c:
```diff
@@ -253,9 +253,9 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext)
-			if (shdr->sh_flags != region->sh_flags) {
-				LOG_ERR("Unsupported section flags for %s (region %d)",
-					name, mem_idx);
+			if ((shdr->sh_flags & 0xf) != (region->sh_flags & 0xf)) {
```
+1

Also, why 0xF? `SHF_WRITE` is 0x1, `SHF_ALLOC` is 0x2, `SHF_EXECINSTR` is 0x4, but I don't see 0x8 being defined anywhere in the spec.

mathieuchopstm commented on include/zephyr/llext/symbol.h:
```diff
@@ -117,18 +117,22 @@ struct llext_symtable {
  */
-#if defined(CONFIG_LLEXT) && defined(LL_EXTENSION_BUILD)
+#if defined(CONFIG_LLEXT)
+#if defined(LL_EXTENSION_BUILD)
```
I think the rationale behind this is to allow exporting symbols from code that may be built in-application or as an LLEXT (as done in `samples/subsys/llext/modules`): https://github.com/zephyrproject-rtos/zephyr/blob/143b14bb4fb779437fdeb8369e42fbb81db4a5fd/samples/subsys/llext/modules/src/hello_world_ext.c#L17-L27

IMO, we should rename `EXPORT_SYMBOL` to `EXPORT_BUILTIN_SYMBOL`, and rename the existing `LL_EXTENSION_SYMBOL` to `EXPORT_EXTENSION_SYMBOL`; then, we would have a single e.g. `LLEXT_SYMBOL_EXPORT` macro that would be used everywhere and could select between the two depending on the type of build.
(For Zephyr code, this would always turn into `EXPORT_BUILTIN_SYMBOL`, unless someone starts having fun with this to make the kernel modular :p)

```c
#if defined(CONFIG_LLEXT)
#define EXPORT_BUILTIN_SYMBOL(x)  ... /* this should select between SLID and non-SLID but you get the idea */
#define EXPORT_EXTENSION_SYMBOL(x) ... /* current LL_EXTENSION_SYMBOL here */
#else
/* stub defines */
#define EXPORT_BUILTIN_SYMBOL(x)
#define EXPORT_EXTENSION_SYMBOL(x)
#endif

#if defined(LL_EXTENSION_BUILD)
#define LLEXT_SYMBOL_EXPORT(x) EXPORT_EXTENSION_SYMBOL(x)
#else
#define LLEXT_SYMBOL_EXPORT(x) EXPORT_BUILTIN_SYMBOL(x)
#endif
```

(sidenote: LL_EXTENSION_BUILD is lacking documentation, although IDK where this kind of macro should be documented)

teburd replied:
The thing is... EXPORT_SYMBOL can be defined to be whatever it needs to be.

E.g. when building an extension it can be defined to do something different than when being built in the base firmware if needed. I don't think we need two distinct macros here myself.

mathieuchopstm replied:
Maybe I worded it poorly, but this is exactly what I wanted to propose: have a single macro used to export symbols, regardless of where we are exporting from, that would automagically select the proper implementation (i.e., `STRUCT_SECTION_ITERABLE(llext_const_symbol, x ## _sym)` if building in base firmware, `Z_GENERIC_SECTION(".exported_sym")` if building in extension).

Maybe I should have used the `Z_` prefix to make it clear that `EXPORT_BUILTIN_SYMBOLS`/`EXPORT_EXTENSION_SYMBOL` are internal-usage-only macros in my proposal snippet.

lyakh replied:
@teburd agree, just like in the Linux kernel

mathieuchopstm commented on subsys/llext/llext_link.c:
Nit: this sounds prone to race conditions - but the whole subsystem is, anyways, as `struct llext` lacks a lock... This is well outside PR scope though :smile:

mathieuchopstm commented on subsys/llext/llext_link.c:
```diff
@@ -52,6 +52,73 @@ static size_t llext_file_offset(struct llext_loader *ldr, size_t offset)
+	return -ENOENT;
+}
+
+void llext_dependency_remove(struct llext *ext)
```
Nit: This name sounds like a poor choice to me; something like `llext_release_dependencies`/`llext_detach_from_dependencies`/... (emphasis on **cies**, i.e., make it clear this drops refcount of *all extensions we depend on*) would be more appropriate IMO.

lyakh replied:
@mathieuchopstm I prefer naming from objects to verbs, i.e. `irq_disable()` not `disable_irq()`. And singular over plural in most cases. So, the choice was deliberate, but if we want to emphasise, that it's removing all dependencies, we could make it `llext_dependency_remove_all()` or similar. And "remove" is chosen to counteract `llext_dependency_add()`. But we could use "release" too, I guess

mathieuchopstm replied:
I see your point. +1 on `llext_dependency_remove_all()`.

mathieuchopstm reviewed (commented):
Neat piece of work!

A few comments from me, but otherwise looks pretty solid:

pillo79 reviewed (commented):
Agree with all comments, I'm late to the party :slightly_smiling_face: 
Otherwise LGTM as well! :+1: Thanks!

lyakh committed:
llext: make EXPORT_SYMBOL() universal

Currently LLEXT has two ways to export symbols:
EXPORT_SYMBOL() to exort symbols from the main firmware to extensions
LL_EXTENSION_SYMBOL() to export symbols from extensions
but it is possible to write code, that needs to export symbols
regardless of whether it is built into the main image or an
extension. And in fact there's no real need in two distinct macros -
we can use one for both cases and recognise at build time which
direction exporting is happening. This patch extends EXPORT_SYMBOL()
to also work from extensions.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: use EXPORT_SYMBOL() universally

Replace LL_EXTENSION_SYMBOL() with EXPORT_SYMBOL() in all tests and
samples.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

pillo79 commented on subsys/llext/llext_link.c:
```diff
@@ -328,6 +389,16 @@ int llext_link(struct llext_loader *ldr, struct llext *ext, bool do_local)
 				link_addr = (uintptr_t)llext_find_sym(NULL,
 					SYM_NAME_OR_SLID(name, sym.st_value));
 
+				if (link_addr == 0) {
+					/* Try loaded tables */
+					struct llext *dep;
+
+					link_addr = (uintptr_t)llext_find_extension_sym(name, &dep);
+					if (link_addr) {
+						llext_dependency_add(ext, dep);
+					}
+				}
+
```
Very minor nit - if you ever have to refactor anything: this bit of code is added in the last commit, which only deals with dependency tracking. Similar to what happens for Xtensa, it could be split between the commit that introduces the actual lookup and the one adding tracking.

pillo79 reviewed (approved):
LGTM, thanks for this! 
Just mentioning a tiny refactor issue, maybe fix it if you ever have to rework commits.

mathieuchopstm reviewed (approved):
LGTM.

Nit: I feel like `llext_find_extension_sym` would be a more fitting name for what we currently call `llext_find_sym`, and vice-versa... just food for thought, changing the API is not in scope for this PR :smile:

lyakh commented:
#77738 adds a (draft but should be already working) test for this new functionality, which will be finalised and released for review and merging once this has been merged

teburd reviewed (approved):
Very nice set of changes

aescolar commented:
Added dnm @lyakh given your initial comment indicating this was a draft, to avoid accidental merge. Please confirm you'd like to have this merged. You (or anybody else) is welcome to remove the dnm.

lyakh commented:
> Added dnm given @lyakh's initial comment indicating this was a draft, to avoid accidental merge. Please confirm you'd like to have this merged. You (or anybody else) is welcome to remove the dnm.

@aescolar sorry, forgot to update the PR description, done now
